### PR TITLE
Oceanwater 879 PanTiltMoveCartesian

### PR DESCRIPTION
**Pull request by AstroStucky**
## Linked Issues:
| EPIC ⚡ | [OCEANWATER-933](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-933) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-879](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-879) |

## Summary of Changes
* **PanTiltMoveCartesian** action added
* Antenna move logic collected into a mixin shared by both antenna move actions.

## Tests
Ensure your **ow_autonomy** is on the latest commit of the epic branch, [OCEANWATER-933-command-unification-epic](https://github.com/nasa/ow_simulator/tree/OCEANWATER-933-command-unification-epic), then rebuild. 

Launch atacama_y1a along with an RQT GUI. 

### Help Message
Check that the language makes sense 
```
rosrun ow_lander pan_tilt_move_cartesian.py --help
```

### Fail to look at stuff
Request an impossible view position. Since pan has full 360 coverage, I believe tilt is the only value that may be unreachable. 
```
rosrun ow_lander pan_tilt_move_cartesian.py -z -5
```

The antenna will not move, and you will see the following output
```
PanTiltMoveCartesian action started
PanTiltMoveCartesian: Aborted - Requested tilt 1.5644910165286037 is not within allowed limits.
PanTiltMoveCartesian action complete
```

### Look at Stuff
To verify mast pose, we will use the images from both cameras. In RQT, open a second Image View by clicking **Plugins > Visualization > Image View**. In the second Image View window, select `/StereoCamera/right/image_raw` from the dropdown list.  Perform the following steps and confirm the left/right camera images agree with the ones I have provided.

Each command should result in the following output from the action
```
PanTiltMoveCartesian action started
PanTiltMoveCartesian: Succeded - Reached commanded pan/tilt values
PanTiltMoveCartesian action complete
```

#### 1. Unstow arm and look at it
```
rosrun ow_lander arm_unstow.py
rosrun ow_lander pan_tilt_move_cartesian.py -f 1
rosrun ow_lander camera_capture_action_client.py
```
Screenshot_20230120_173608

#### 2. Look at the reddish mound in front of lander

```
rosrun ow_lander pan_tilt_move_cartesian.py -x 1.46 -y -0.4 -z -0.05
rosrun ow_lander camera_capture_action_client.py
```
Screenshot_20230120_173723

#### 3. Send arm to some coordinate, then look at that coordinate with **PanTiltMoveCartesian**
```
rosrun ow_lander arm_move_cartesian.py -x 0.5 -y -0.5 -z 1.5
rosrun ow_lander pan_tilt_move_cartesian.py -x 0.5 -y -0.5 -z 1.5
rosrun ow_lander camera_capture_action_client.py
```
Screenshot_20230120_174426

This may look wrong, but we must keep in mind three things: 
**A**) The tip of the scoop's blade is the subject of focus because that is what is posed by **ArmMoveCartesian**
**B**) Neither camera will show the scoop tip in the center of their image, but rather slightly to the right of center in the left cam and slightly to the left of center in the right cam
**C**) The distance the subject is off from the center in both images increases the closer the subject is to the cameras. 

#### 4. Pretend to discard sample and image where we would expect sample to land
```
rosrun ow_lander discard_action_client.py
rosrun ow_lander pan_tilt_move_cartesian.py -f 1 -x 0.7
rosrun ow_lander camera_capture_action_client.py
```
Screenshot_20230120_175527

#### 5. Take in the view
```
rosrun ow_lander pan_tilt_move_cartesian.py -x -125800 -y -81170 -z 80533
rosrun ow_lander camera_capture_action_client.py
```
Screenshot_20230120_175803

### CODE REVIEW

**Comment by AstroStucky** [comment]
Epic branch merged. Tests still work.

**Comment by kmdalal** [comment]
Should note that `ow_autonomy` needs to be on the OW-333 epic branch.   Also this branch needed the same epic branch merged in order to build without errors.

**Comment by kmdalal** [comment]
The `pan_tilt_move_cartesian.py` script is missing -- are you sure it was pushed?

**Comment by kmdalal** [comment]
> Should note that `ow_autonomy` needs to be on the OW-333 epic branch. Also this branch needed the same epic branch merged in order to build without errors.

Never mind the second sentence.  I switched to this branch without having done a pull recently.

**Comment by kmdalal** [comment]
> The `pan_tilt_move_cartesian.py` script is missing -- are you sure it was pushed?

I just pulled and it came in.

**Comment by kmdalal** [comment]
The 'fail to look at stuff' test gave me this error:
```
[INFO]: PanTiltMoveCartesian action started
[INFO]: PanTiltMoveCartesian: Succeeded - Reached commanded pan/tilt values
[ERROR]: pan_position is not an attribute of PanTiltMoveCartesianResult; an empty action result will be published
[INFO]: PanTiltMoveCartesian action complete
```

**Comment by AstroStucky** [comment]
> Every call to the client script gives me this error:
> 
> ```
> [ERROR]: pan_position is not an attribute of PanTiltMoveCartesianResult; an empty action result will be published
> ```

I have not been able to replicate this output, and I cannot find anywhere in the code that might cause it.

**Comment by AstroStucky** [comment]
I am stumped by your problem, Mike. Could you please try again following a clean and build just to be absolutely sure there aren't branches getting crossed?

**Comment by kmdalal** [comment]
> I am stumped by your problem, Mike. Could you please try again following a clean and build just to be absolutely sure there aren't branches getting crossed?

I deleted and re-checked out the branch, built clean, and the problem is gone.  Not sure why I keep having these kind of issues!

**Review by kmdalal — CHANGES_REQUESTED** [review]
All tests passed, and nicely selected.  Code review light but everything looked clear and readable.   The only thing I found missing is the action result and feedback, as specified in the Google doc and also implemented by OWLAT.

File: owl_msgs/action/PanTiltMoveCartesian.action [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# goal definition
+int32 frame
+geometry_msgs/Point point   # position in space antenna mast will look at
+---
+# result
+---
+# feedback
```
Comment (kmdalal) [inline owl_msgs/action/PanTiltMoveCartesian.action:7]: pan and tilt position needed for both result and feedback

File: ow_lander/scripts/pan_tilt_move_cartesian.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+from ow_lander import node_helper
+from ow_lander import constants
+
+import argparse
+
+from geometry_msgs.msg import Point
+
+parser = argparse.ArgumentParser(
+  formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+  description="Moves the antenna mast to look position in space.")
```
Comment (mogumbo) [inline ow_lander/scripts/pan_tilt_move_cartesian.py:17]: look "at a" position in space

**Review by mogumbo — CHANGES_REQUESTED** [review]

File: ow_lander/src/ow_lander/actions.py [hunk 3]
```diff
@@ -706,99 +706,92 @@ def execute_action(self, _goal):
...
+    # lever arm can be approximated from only their lengths.
+    a = math.acos(l / math3d.norm(tilt_to_lookat))
+    # compute the angle between tilt_to_lookat and the X-Y plane
+    b = math.atan2(tilt_to_lookat.z,
+                   math.sqrt(tilt_to_lookat.x**2 + tilt_to_lookat.y**2))
+    # The sum of a and b gives the angle between the desired camera lever arm
+    # vector and the x-y plane.
+    # Antenna tilt is measured from the +z axis, so a pi/2 is subtracted.
+    # Finally, tilt rotates in reverse of the unit circle, so we multiple the
+    # the result by zero.
```
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_lander/src/ow_lander/actions.py:786]: yes 😄

File: owl_msgs/action/PanTiltMoveCartesian.action (hunk 1 above)
Comment (AstroStucky) [reply owl_msgs/action/PanTiltMoveCartesian.action:7]: That is only the case for **PanTitleMoveJoint**, but the unification document has result and feedback empty for this action.

File: ow_lander/scripts/pan_tilt_move_cartesian.py (hunk 2 above)
Comment (AstroStucky) [reply ow_lander/scripts/pan_tilt_move_cartesian.py:17]:
```suggestion
  description="Moves the antenna mast to look at a position in space.")
```

File: ow_lander/src/ow_lander/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_lander/src/ow_lander/actions.py:786]:
```suggestion
    # the result by -1.
```

[the remark this answers is not in the document]
Comment (mogumbo) [reply ow_lander/src/ow_lander/actions.py]: Pretty good, but it still says "multiple the the result" instead of "multiply the result"

**Review by mogumbo — APPROVED** [review]

File: owl_msgs/action/PanTiltMoveCartesian.action (hunk 1 above)
Comment (kmdalal) [reply owl_msgs/action/PanTiltMoveCartesian.action:7]: Per our discussion, the released OWLAT file is actually wrong, and your action file matches the unification document.  Let's stick with this action file.

**Review by kmdalal — APPROVED** [review]
We have resolved the action file issue.  Nice work!

File: ow_lander/src/ow_lander/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_lander/src/ow_lander/actions.py]: Whoops, will fix that real quick

File: ow_lander/src/ow_lander/actions.py [hunk 4]
```diff
@@ -706,99 +706,92 @@ def execute_action(self, _goal):
...
+    # right triangle. Therefore, the angle between tilt_to_lookat and the camera
+    # lever arm can be approximated from only their lengths.
+    a = math.acos(l / math3d.norm(tilt_to_lookat))
+    # compute the angle between tilt_to_lookat and the X-Y plane
+    b = math.atan2(tilt_to_lookat.z,
+                   math.sqrt(tilt_to_lookat.x**2 + tilt_to_lookat.y**2))
+    # The sum of a and b gives the angle between the desired camera lever arm
+    # vector and the x-y plane.
+    # Antenna tilt is measured from the +z axis, so a pi/2 is subtracted.
+    # Finally, tilt rotates in reverse of the unit circle, so we multiple the
```
Comment (AstroStucky) [inline ow_lander/src/ow_lander/actions.py:785]:
```suggestion
    # Finally, tilt rotates in reverse of the unit circle, so we multiply the
```

**Pull request merged**